pkg/expression/builtin_control.go

```diff
@@ -255,7 +255,8 @@ func InferType4ControlFuncs(ctx BuildContext, funcName string, args ...Expressio
 	nullFields := make([]*types.FieldType, 0, argsNum)
 	notNullFields := make([]*types.FieldType, 0, argsNum)
 	for i := range args {
-		if args[i].GetType(ctx.GetEvalCtx()).GetType() == mysql.TypeNull {
+		ci, ok := args[i].(*Constant)
+		if args[i].GetType(ctx.GetEvalCtx()).GetType() == mysql.TypeNull || (ok && ci.Value.Kind() == types.KindNull) {
```

Comment:
I've a little concern about this change, sometimes the null constant do have its own type, for example, user can use `cast(null as xxx)` to create a null constant with specific type, but in this pr, it ignores the null constant type, which I think will cause some inconsistencies with MySQL behavior.
For example, consider the following sql
`select case when 0 then cast(null as char(100)) else 1 end;`
The return type is `char` before this pr, and it will return `bigint` after this pr, and in mysql, the return type is `char`.

Reply:
And I will hold this pr in case it would be merged accidentally.